packages/hardhat-core/test/internal/hardhat-network/jsonrpc/client.ts

```diff
@@ -23,402 +25,450 @@ type FakeProvider = Pick<HttpProvider, "url" | "sendBatch"> & {
-            true
-          );
-          assert.isTrue(
-            block?.transactions.every(
-              (tx: Buffer | RpcTransaction) => !(tx instanceof Buffer)
-            )
+        it("Reads from disk if available, not making any request a request", async function () {
+          await makeCall();
+          assert.isTrue((fakeProvider.request as sinon.SinonStub).calledOnce);
+
+          clientWithFakeProvider = new JsonRpcClient(
+            fakeProvider as any,
+            1,
+            123,
+            3,
+            this.tmpDir
           );
-        });
 
-        it("returns null for non-existent block", async () => {
-          const block = await client.getBlockByHash(randomHashBuffer(), true);
-          assert.isNull(block);
+          await makeCall();
+          assert.isTrue((fakeProvider.request as sinon.SinonStub).calledOnce);
         });
       });
+    });
 
-      describe("eth_getCode", () => {
-        it("can fetch code of an existing contract", async () => {
-          const code = await client.getCode(DAI_ADDRESS, "latest");
-          assert.notEqual(bufferToHex(code), "0x");
-        });
+    describe("Retry on Infura's error", () => {
+      const fakeInfuraUrl = "http://infura.com";
 
-        it("can fetch empty code of a non-existent contract", async () => {
-          const code = await client.getCode(EMPTY_ACCOUNT_ADDRESS, "latest");
-          assert.equal(bufferToHex(code), "0x");
-        });
+      const response =
+        "0x00000000000000000000000000000000000000000067bafa8fb7228f04ffa792";
+
+      it("makes a retry on the 'header not found' error", async () => {
+        const fakeProvider: FakeProvider = {
+          url: fakeInfuraUrl,
+          request: sinon
+            .stub()
+            .onFirstCall()
+            .rejects(new Error("header not found"))
+            .onSecondCall()
+            .resolves(response),
+          sendBatch: () => Promise.resolve([]),
+        };
+
+        const clientWithFakeProvider = new JsonRpcClient(
+          fakeProvider as any,
+          1,
+          123,
+          3
+        );
+        const value = await clientWithFakeProvider.getStorageAt(
+          DAI_ADDRESS,
+          DAI_TOTAL_SUPPLY_STORAGE_POSITION,
+          new BN(120)
+        );
+        assert.equal((fakeProvider.request as sinon.SinonStub).callCount, 2);
+        assert.isTrue(value.equals(toBuffer(response)));
       });
 
-      describe("eth_getStorageAt", () => {
-        it("can fetch value from storage of an existing contract", async () => {
-          const totalSupply = await client.getStorageAt(
+      it("does not retry more than once", async () => {
+        const fakeProvider: FakeProvider = {
+          url: fakeInfuraUrl,
+          request: sinon
+            .stub()
+            .onFirstCall()
+            .rejects(new Error("header not found"))
+            .onSecondCall()
+            .rejects(new Error("header not found"))
+            .onCall(4)
```

Comment:
Shouldn't this be `onCall(3)`?

Reply:
Yes, this is why I said I hate this api lol